feat: limit inbound queue

mhofman opened the pull request:
closes: #6306 #6313 #6314

closes: #6381 (adopts stacked commit)

## Description

This updates the go->swingset BlockingSend path to include a JSON parsed result, and uses that new path to communicate during begin block a number of messages go should add to the action queue. This allowed value is currently computed from a new inbound queue max block param. It is enforced by a new ante handler by rejecting transactions that would put the action queue over the number of new allowed messages.

A separate mempool value is computed by using half the max inbound size. This provides some buffer to avoid including transactions in a block that would then get rejected.

To deal with hangover inconsistency, the allowed queue sizes are stored in the host section of the swingstore, and replayed at begin block if the computed height logic detects an inconsistency (checks factored out of end block).

### Security Considerations

Cosmos may end up queueing more than the allowed messages since only external transactions going through the ante handler are subject to the check, and internally generated messages such as vbank updates are processed after. This is fine since these are usually low overhead, and will simply be accounted for at the next block anyway.

### Documentation Considerations

I picked a sensible default of 1000 for the max inbound queue size, but like other params, this is configurable by governance.

### Testing Considerations

Tested locally by setting the max inbound queue size to 5, and a loadgen with 32 tasks in parallel. The solo client did cause the allowed inbound to drop to 0, and triggered "mempool full" errors when trying to send a message to the chain.

mhofman commented:
As usual, best reviewed commit-by-commit.

@michaelfig for primary review since both @JimLarson and I worked on this PR
@warner for the architecture of the whole thing

warner commented on golang/cosmos/ante/inbound.go:
```diff
@@ -0,0 +1,73 @@
+			*swingtypes.MsgWalletSpendAction:
+			if inboundsAllowed == -1 {
+				state := ia.sk.GetState(ctx)
+				allowed, found := swingtypes.QueueSizeEntry(state.QueueAllowed, "inbound")
```
Does `allowed` mean the maximum (1000), or the remaining space? The `allowed - actions` below suggests that it means the maximum, but I'm not sure that's how it was used over in the JS side.

warner commented on golang/cosmos/ante/inbound.go:
```diff
@@ -0,0 +1,73 @@
+				}
+			}
+			if inboundsAllowed > 0 {
+				inboundsAllowed--
```
This looks like it's local to this function/method, and it looks like it's called with a single txn (which might contain multiple messages). Do ante handlers get run for all txns before any of them are processed? Or does it do `ante handler, process txn` as a pair, before moving on to the next one? If it's the former, then a local (ephemeral) `inboundsAllowed` would reset for each txn, and allow in too many.

JimLarson replied:
I'm taking a look, but in the worst case, we let too many txns in, and then in the *next* block we reject them all, giving us a little hysteresis, which isn't necessarily a bad thing.

JimLarson replied:
Confirmed that the ante handler runs at `DeliverTx` time, so we're seeing the state mutated by previous Txs in the block. So we're going to reject the Tx if it has enough swingset Msgs to overflow the inbound queue (counting the pending action queue).

warner commented on golang/cosmos/ante/inbound.go:
```diff
@@ -0,0 +1,73 @@
+			*swingtypes.MsgProvision,
+			*swingtypes.MsgWalletAction,
+			*swingtypes.MsgWalletSpendAction:
+			if inboundsAllowed == -1 {
```
ah, ok so this queries the state at most once per txn (not once per msg), and even then only if there were any swingset messages in the txn, nice

JimLarson replied:
It even breaks off from iterating through all the Msgs once it discovers that it goes over the limit. The could could be a little more clear if I just summed up all the swingset Msgs in the Tx, rather than computing mid-loop, but it's not too awful as it is.

warner commented on golang/cosmos/x/swingset/keeper/keeper.go:
```diff
@@ -117,6 +120,22 @@ func (k Keeper) actionQueueIndex(ctx sdk.Context, name string) (uint64, error) {
+		return 0, err
+	}
+	size := tail - head
+	if size > math.MaxInt32 {
```
yeah let's hope our `inboundQueue` never gets this big :)

warner reviewed (approved):
Looks pretty good to me, although the names sometimes felt confusing (but if they're consistent, across both JS and Go, that's the most important thing, I'll read it again in the morning and maybe it'll speak to me differently). The only potential logic bug I'm worried about is that `allowed` seems to mean "remaining space" (`inboundQueueMax - inboundQueueSize`) in launch-chain.js, and .. oh, ok the Go side is looking at the *action* queue, not the *inbound* queue, so `allowed` is in fact the remaining space (in the inbound queue), and the golang side is doing `actions = ia.sk.ActionQueueLength` to measure the current action queue, so `allowed - actions` is the remaining space in the *action* queue, and hence what matters when `ante/inbound.go` is deciding whether to accept or reject each msg.

Ok, I'm not sure why I got confused by that, but it seems right.

mhofman reviewed (commented):
I'll add some comments around the confusing bits and rewrite the computed height check.
Happy to bikeshed the "allowed" naming, but I'd like alternatives to "remaining" which is too local to the JS side.

mhofman commented on packages/cosmic-swingset/src/launch-chain.js:
```diff
@@ -559,6 +578,23 @@ export async function launch({
     return p;
   }
 
+  function checkExpectedBlockHeight(blockHeight) {
```
So looking into this more I think I understand the current check better, and I'm having a hard time finding a better expression of it. Here are 2 constraints missing above:
- The bootstrap block (assumed blockHeight=0) does not get a commit. The first commit is after the first block.
- The first block may not be block 1, but instead be an arbitrary number.

That means getting a `blockHeight != computedHeight - 1` is actually valid if `computedHeight === 0`, hence the first part of the check.

mhofman commented:
Sorry for the churn, I messed up the rebase to target master. Here is the list of commits addressing feedback since the initial push last night: https://github.com/Agoric/agoric-sdk/pull/6381/files/ea3ee50b0efa1884284d8e4647ce8b321620ac4d..4e1221767a2e01b091a8a3d198be975505ee37ca

warner commented on packages/cosmic-swingset/src/launch-chain.js:
```diff
@@ -586,21 +586,26 @@ export async function launch({
-        throw decohered;
-      }
+  function blockNeedsExecution(blockHeight) {
+    if (blockHeight > 0 && savedHeight === blockHeight) {
```
wait, shouldn't that be `if (savedHeight > 0` ?

Would it be correct for this to read:

```js
if (savedHeight === 0) {
  // 0 is the default we use when the DB is empty, so either we haven't executed any blocks, or we've executed the bootstrap block but no others. The first non-bootstrap block can have an arbitrary height (the chain may not start at 1), so we can't tell if it is a replay or not, so assume the block needs to be executed
  return true;
}

if (blockHeight === savedHeight + 1) {
  // execute the next block
  return true;
}

if (blockHeight === savedHeight) {
  // we have already committed this block, so "replay" by not executing (but returning all the results from the last time)
  return false;
}

// we're being asked to rewind by more than one block, or execute something more than one block in the future, neither of which we can accomodate
decohered = ...
throw decohered;
```

warner commented:
Now that `release-pismo` has been updated/merged with trunk, please rebase on top of current trunk (7e071a7) before landing.

warner reviewed (approved):
New changes look good, but consider the rewrite; I haven't confirmed that the existing version (`if (blockHeight > 0 && savedHeight === blockHeight) {`) is correct.

mhofman commented:
I will rebase and fixup once master settles to avoid re-running CI multiple times.

I'd still like to get an approval from @michaelfig, and a +1 from @JimLarson to make sure the comments I added to the ante handler are sane.

JimLarson commented on golang/cosmos/ante/inbound.go:
```diff
@@ -0,0 +1,102 @@
+					entry = QueueInboundMempool
+				}
+				allowed, found := swingtypes.QueueSizeEntry(state.QueueAllowed, entry)
+				if !found && simulate {
```
It looks like we misunderstood the `simulate` parameter - it does not mean it's running for CheckTx - i.e. it's not the signal that we're running for mempool admission. It's not clear what the signal is for mempool context. I'm continuing to dig.

JimLarson reviewed (approved):
Some flaws, but not critical, so LGTM.

Issues identified in my comments are fixed in #6392

mhofman commented on golang/cosmos/ante/inbound.go:
```diff
@@ -0,0 +1,102 @@
+				if !found && simulate {
+					entry = QueueInbound
+					allowed, found = swingtypes.QueueSizeEntry(state.QueueAllowed, entry)
+					allowed /= 2
```
Fair. Closing shut is safer

mhofman reviewed (commented):
> Issues identified in my comments are fixed in https://github.com/Agoric/agoric-sdk/pull/6392

Adopting that last stacked commit. Will now rebase everything and merge.

Fixup: https://github.com/Agoric/agoric-sdk/compare/ac89ecd7868f2dfdee13052e9a06e2dc555e7279..46c2c935c820acf5b7563defb96aef63fe0508f2

michaelfig commented on golang/cosmos/x/swingset/keeper/keeper.go:
```diff
@@ -29,6 +31,10 @@ const (
 	StoragePathBundles      = "bundles"
 )
 
+const MaxUint53 = 9007199254740991 // Number.MAX_SAFE_INTEGER = 2**53 - 1
```
```suggestion
const MaxUint53 = 2<<52 - 1 // Number.MAX_SAFE_INTEGER = 2**53 - 1
```

mhofman replied:
Oh I had tried something like that, and couldn't figure it out in Go

JimLarson replied:
Should fix either the code or the comment to settle on 52 vs 53.

JimLarson replied:
And the right answer: can exactly store 2**53 in a float, 2**53 + 1 is the first integer that can't be represented in a double float.

mhofman replied:
The code and the comment are both correct: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER

mhofman replied:
To clarify we can rewrite it as `1<<53 - 1` next time we touch this.

michaelfig commented on golang/cosmos/x/swingset/keeper/keeper.go:
```diff
@@ -78,22 +84,22 @@ func NewKeeper(
+	}
+
+	// JS uses IEEE 754 floats so avoid overflowing integers
+	if tail == MaxUint53 {
```
Could be more defensive.

```suggestion
	if tail >= MaxUint53 {
```

mhofman replied:
Arguably we're already busted then ;)

mhofman commented:
Forced push without changes in hopes to nudge CI

michaelfig commented on packages/cosmic-swingset/src/launch-chain.js:
```diff
@@ -300,6 +301,23 @@ export async function launch({
+  function computeQueueAllowed(_blockHeight, _blockTime, params) {
+    assert(params.queueMax);
+
+    assert(QueueInbound in params.queueMax);
```
`in` is fine, IMO.

michaelfig reviewed (approved):
LGTM!  Approach is good, and I have just a few minor comments.
